src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs

```diff
@@ -209,19 +210,70 @@ public RequestDelegate CreateMiddleware(RequestDelegate next)
             }
         }
 
+        // Performance optimization: Precompute and cache the key results for each parameter
+        var precomputedKeys = new object?[parameters.Length];
+        for (var i = 1; i < parameters.Length; i++)
+        {
+            var hasServiceKey = TryGetServiceKey(parameters[i], out object? key);
+            precomputedKeys[i] = key;
+        }
+
         return (middleware, context, serviceProvider) =>
         {
             var methodArguments = new object[parameters.Length];
             methodArguments[0] = context;
             for (var i = 1; i < parameters.Length; i++)
             {
-                methodArguments[i] = GetService(serviceProvider, parameters[i].ParameterType, methodInfo.DeclaringType!);
+                var key = precomputedKeys[i];
+                var parameterType = parameters[i].ParameterType;
+                var declaringType = methodInfo.DeclaringType!;
+
+                methodArguments[i] = key == null ? GetService(serviceProvider, parameterType, declaringType) : GetKeyedService(serviceProvider, key, parameterType, declaringType);
             }
 
             return (Task)methodInfo.Invoke(middleware, BindingFlags.DoNotWrapExceptions, binder: null, methodArguments, culture: null)!;
         };
     }
 
+    private static bool TryGetServiceKey(ParameterInfo parameterInfo, out object? key)
+    {
+        if (parameterInfo.CustomAttributes != null)
+        {
+            foreach (var attribute in parameterInfo.GetCustomAttributes(true))
```

Comment:
`parameterInfo.OfType<FromKeyedServicesAttribute>().FirstOrDefault()` might be more straightforward here.

Reply:
@captainsafia, I've taken your feedback into account and changed the method to be more straightforward. Also changed the Inherit parameter from true to false, to meet the ActivatorUtilities from the Microsoft.Extensions.DependencyInjections.Abstractions package.

https://github.com/dotnet/runtime/blob/0540ea54f3ab96ed02e22e242fe9064cbda2a63c/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs#L416

https://github.com/dotnet/runtime/blob/0540ea54f3ab96ed02e22e242fe9064cbda2a63c/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs#L749

https://github.com/dotnet/runtime/blob/0540ea54f3ab96ed02e22e242fe9064cbda2a63c/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs#L532C3-L532C41

Reply:
@captainsafia, @BrennanConroy, the changes within this pull-request target main, and thus the .NET 9.0 version. If I'd want the changes to be available in the latest .NET 8.x.x version as well, what would the appropriate steps for me to take?

Reply:
@captainsafia, @BrennanConroy, any follow-up on my question above?

Reply:
I don't think this is a backport candidate. This change is a feature, not fixing a major bug/regression introduced in 8.0.